# **BƯỚC 1 : ĐỌC DỮ LIỆU**

In [1]:
from google.colab import drive
drive.mount('/content/drive')
import pandas as pd
import plotly.express as px
df = pd.read_csv('/content/drive/My Drive/Colab Notebooks/pandas/ex/Airbnb_Data/Airbnb_Open_Data.csv')
df.head()

Mounted at /content/drive


<ipython-input-1-21fd3bd8c4a3>:5: DtypeWarning: Columns (25) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('/content/drive/My Drive/Colab Notebooks/pandas/ex/Airbnb_Data/Airbnb_Open_Data.csv')


id                                              NAME      host id  \
0  1001254                Clean & quiet apt home by the park  80014485718   
1  1002102                             Skylit Midtown Castle  52335172823   
2  1002403               THE VILLAGE OF HARLEM....NEW YORK !  78829239556   
3  1002755                                               NaN  85098326012   
4  1003689  Entire Apt: Spacious Studio/Loft by central park  92037596077   

  host_identity_verified host name neighbourhood group neighbourhood  \
0            unconfirmed  Madaline            Brooklyn    Kensington   
1               verified     Jenna           Manhattan       Midtown   
2                    NaN     Elise           Manhattan        Harlem   
3            unconfirmed     Garry            Brooklyn  Clinton Hill   
4               verified    Lyndon           Manhattan   East Harlem   

        lat      long        country  ... service fee minimum nights  \
0  40.64749 -73.97237  United States  ...       $193            10.0   
1  40.75362 -73.98377  United States  ...        $28            30.0   
2  40.80902 -73.94190  United States  ...       $124             3.0   
3  40.68514 -73.95976  United States  ...        $74            30.0   
4  40.79851 -73.94399  United States  ...        $41            10.0   

  number of reviews last review  reviews per month review rate number  \
0               9.0  10/19/2021               0.21                4.0   
1              45.0   5/21/2022               0.38                4.0   
2               0.0         NaN                NaN                5.0   
3             270.0    7/5/2019               4.64                4.0   
4               9.0  11/19/2018               0.10                3.0   

  calculated host listings count  availability 365  \
0                            6.0             286.0   
1                            2.0             228.0   
2                            1.0             352.0   
3                            1.0             322.0   
4                            1.0             289.0   

                                         house_rules license  
0  Clean up and treat the home the way you'd like...     NaN  
1  Pet friendly but please confirm with me if the...     NaN  
2  I encourage you to use my kitchen, cooking and...     NaN  
3                                                NaN     NaN  
4  Please no smoking in the house, porch or on th...     NaN  

[5 rows x 26 columns]

In [ ]:
# số cột k hiển thị hết do bị tràn, tên cột hoa thường lẫn lộn, ta cần setup lại:
pd.set_option('display.max_columns',500)   # thiết lập số lượng tối đa pd in ra
pd.set_option('display.max_row',500)
print(df.shape) #cung cấp thông tin về kích thước của dữ liệu bạn đang làm việc.
df.head()

(102599, 26)


id                                              NAME      host id  \
0  1001254                Clean & quiet apt home by the park  80014485718   
1  1002102                             Skylit Midtown Castle  52335172823   
2  1002403               THE VILLAGE OF HARLEM....NEW YORK !  78829239556   
3  1002755                                               NaN  85098326012   
4  1003689  Entire Apt: Spacious Studio/Loft by central park  92037596077   

  host_identity_verified host name neighbourhood group neighbourhood  \
0            unconfirmed  Madaline            Brooklyn    Kensington   
1               verified     Jenna           Manhattan       Midtown   
2                    NaN     Elise           Manhattan        Harlem   
3            unconfirmed     Garry            Brooklyn  Clinton Hill   
4               verified    Lyndon           Manhattan   East Harlem   

        lat      long        country country code instant_bookable  \
0  40.64749 -73.97237  United States           US            False   
1  40.75362 -73.98377  United States           US            False   
2  40.80902 -73.94190  United States           US             True   
3  40.68514 -73.95976  United States           US             True   
4  40.79851 -73.94399  United States           US            False   

  cancellation_policy        room type  Construction year  price service fee  \
0              strict     Private room             2020.0  $966        $193    
1            moderate  Entire home/apt             2007.0  $142         $28    
2            flexible     Private room             2005.0  $620        $124    
3            moderate  Entire home/apt             2005.0  $368         $74    
4            moderate  Entire home/apt             2009.0  $204         $41    

   minimum nights  number of reviews last review  reviews per month  \
0            10.0                9.0  10/19/2021               0.21   
1            30.0               45.0   5/21/2022               0.38   
2             3.0                0.0         NaN                NaN   
3            30.0              270.0    7/5/2019               4.64   
4            10.0                9.0  11/19/2018               0.10   

   review rate number  calculated host listings count  availability 365  \
0                 4.0                             6.0             286.0   
1                 4.0                             2.0             228.0   
2                 5.0                             1.0             352.0   
3                 4.0                             1.0             322.0   
4                 3.0                             1.0             289.0   

                                         house_rules license  
0  Clean up and treat the home the way you'd like...     NaN  
1  Pet friendly but please confirm with me if the...     NaN  
2  I encourage you to use my kitchen, cooking and...     NaN  
3                                                NaN     NaN  
4  Please no smoking in the house, porch or on th...     NaN

# **BƯỚC 2 : PHÂN TÍCH DỮ LIỆU, TÌM RA DỮ LIỆU NONE, ĐIỂM KHÔNG HỢP LỆ**

In [ ]:
#xem kiểu dữ liệu từ cột
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102599 entries, 0 to 102598
Data columns (total 26 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   id                              102599 non-null  int64  
 1   NAME                            102349 non-null  object 
 2   host id                         102599 non-null  int64  
 3   host_identity_verified          102310 non-null  object 
 4   host name                       102193 non-null  object 
 5   neighbourhood group             102570 non-null  object 
 6   neighbourhood                   102583 non-null  object 
 7   lat                             102591 non-null  float64
 8   long                            102591 non-null  float64
 9   country                         102067 non-null  object 
 10  country code                    102468 non-null  object 
 11  instant_bookable                102494 non-null  object 
 12  cancellation_pol

XIN PHÍA ĐẦU DỮ LIỆU, Ý NGHĨA CỦA TỪNG CỘT

| Tên     | Mô tả |
| ----------- | ----------- |
| host_id      | Mã định danh của chủ nhà, khi bạn đăng kí cho thuê trên Airbnb, bên đó sẽ gán cho bạn 1 cái mã       |
| country   | Quốc gia của căn nhà        |
| host_identity_verified   | Chủ nhà có được xác thực hay chưa, hay là địa chỉ kinh doanh ảo       |
| minimum nights   | Số đêm tối thiểu mà chủ nhà cho phép đặt       |
| license   | Có thể là giấy phép kinh doanh      |
| last review   | Ngày được review cuối cùng      |

In [ ]:
df.describe() #mô tả về df theo các thông số

id       host id            lat           long  \
count  1.025990e+05  1.025990e+05  102591.000000  102591.000000   
mean   2.914623e+07  4.925411e+10      40.728094     -73.949644   
std    1.625751e+07  2.853900e+10       0.055857       0.049521   
min    1.001254e+06  1.236005e+08      40.499790     -74.249840   
25%    1.508581e+07  2.458333e+10      40.688740     -73.982580   
50%    2.913660e+07  4.911774e+10      40.722290     -73.954440   
75%    4.320120e+07  7.399650e+10      40.762760     -73.932350   
max    5.736742e+07  9.876313e+10      40.916970     -73.705220   

       Construction year  minimum nights  number of reviews  \
count      102385.000000   102190.000000      102416.000000   
mean         2012.487464        8.135845          27.483743   
std             5.765556       30.553781          49.508954   
min          2003.000000    -1223.000000           0.000000   
25%          2007.000000        2.000000           1.000000   
50%          2012.000000        3.000000           7.000000   
75%          2017.000000        5.000000          30.000000   
max          2022.000000     5645.000000        1024.000000   

       reviews per month  review rate number  calculated host listings count  \
count       86720.000000       102273.000000                   102280.000000   
mean            1.374022            3.279106                        7.936605   
std             1.746621            1.284657                       32.218780   
min             0.010000            1.000000                        1.000000   
25%             0.220000            2.000000                        1.000000   
50%             0.740000            3.000000                        1.000000   
75%             2.000000            4.000000                        2.000000   
max            90.000000            5.000000                      332.000000   

       availability 365  
count     102151.000000  
mean         141.133254  
std          135.435024  
min          -10.000000  
25%            3.000000  
50%           96.000000  
75%          269.000000  
max         3677.000000

In [ ]:
# Không nhìn hết được thì đảo chiều
df.describe().transpose()

count          mean           std  \
id                              102599.0  2.914623e+07  1.625751e+07   
host id                         102599.0  4.925411e+10  2.853900e+10   
lat                             102591.0  4.072809e+01  5.585652e-02   
long                            102591.0 -7.394964e+01  4.952126e-02   
Construction year               102385.0  2.012487e+03  5.765556e+00   
minimum nights                  102190.0  8.135845e+00  3.055378e+01   
number of reviews               102416.0  2.748374e+01  4.950895e+01   
reviews per month                86720.0  1.374022e+00  1.746621e+00   
review rate number              102273.0  3.279106e+00  1.284657e+00   
calculated host listings count  102280.0  7.936605e+00  3.221878e+01   
availability 365                102151.0  1.411333e+02  1.354350e+02   

                                         min           25%           50%  \
id                              1.001254e+06  1.508581e+07  2.913660e+07   
host id                         1.236005e+08  2.458333e+10  4.911774e+10   
lat                             4.049979e+01  4.068874e+01  4.072229e+01   
long                           -7.424984e+01 -7.398258e+01 -7.395444e+01   
Construction year               2.003000e+03  2.007000e+03  2.012000e+03   
minimum nights                 -1.223000e+03  2.000000e+00  3.000000e+00   
number of reviews               0.000000e+00  1.000000e+00  7.000000e+00   
reviews per month               1.000000e-02  2.200000e-01  7.400000e-01   
review rate number              1.000000e+00  2.000000e+00  3.000000e+00   
calculated host listings count  1.000000e+00  1.000000e+00  1.000000e+00   
availability 365               -1.000000e+01  3.000000e+00  9.600000e+01   

                                         75%           max  
id                              4.320120e+07  5.736742e+07  
host id                         7.399650e+10  9.876313e+10  
lat                             4.076276e+01  4.091697e+01  
long                           -7.393235e+01 -7.370522e+01  
Construction year               2.017000e+03  2.022000e+03  
minimum nights                  5.000000e+00  5.645000e+03  
number of reviews               3.000000e+01  1.024000e+03  
reviews per month               2.000000e+00  9.000000e+01  
review rate number              4.000000e+00  5.000000e+00  
calculated host listings count  2.000000e+00  3.320000e+02  
availability 365                2.690000e+02  3.677000e+03

Các dữ liệu thống kê bên trên cho mình thấy sự không hợp lý ở một số cột dữ liệu, ví dụ **minimum nights** sao lại có số âm,dùng cái bảng này để đánh giá ý nghĩa các cột số

Để rõ hơn số dòng dữ liệu Null thì:

In [ ]:
df.isnull().sum().sort_values(ascending = False) ## Đếm số lượng giá trị bị thiếu và sắp xếp giảm dần

license                           102597
house_rules                        52131
last review                        15893
reviews per month                  15879
country                              532
availability 365                     448
minimum nights                       409
host name                            406
review rate number                   326
calculated host listings count       319
host_identity_verified               289
service fee                          273
NAME                                 250
price                                247
Construction year                    214
number of reviews                    183
country code                         131
instant_bookable                     105
cancellation_policy                   76
neighbourhood group                   29
neighbourhood                         16
long                                   8
lat                                    8
id                                     0
host id         

# **BƯỚC 3 : LÀM SẠCH DỮ LIỆU**

### Qua dữ liệu làm sạch :


1.   Xử lý cột `last review`
2.   `host_identity_verified` có giá trị Null => mình sẽ chuyển chúng sang `unconfirmed` hết
3.   Xử lý `minimum nights,availability 365`(có các gtri là số âm), `min=-1223` `max=5645`
4.   Xử lý cột `price`


**Giá trị trung vị (median)** là giá trị nằm ở vị trí giữa của một tập hợp dữ liệu


- tính gtri trung vị
      - dãy dữ liệu lẻ : 3,5,7,9,11. Gtri trung vị là số 7, gtri nằm giữa của dãy
      - dãy dữ liệu chẵn : 3,5,7,9. Gtri trung vị là trung bình cộng của 2 gtri nằm giữa dãy là (5+7)/2 = 6
- thường sử dụng trong các tình huống:
      - Khi dữ liệu có các giá trị ngoại lệ (như quá nhiều none, quá nhiều số âm k hợp lệ, các giá trị này có thể cực kỳ lớn hoặc cực kỳ nhỏ so với các giá trị còn lại, và chúng có thể ảnh hưởng đáng kể đến các phân tích thống kê)
      - Khi làm việc với dữ liệu theo thang đo thứ tự
      - Khi dữ liệu không đối xứng
-Khi nào nên sử dụng giá trị trung bình (mean) thay vì trung vị?
      - khi sử dụng toàn bộ thông tin trong dữ liệu
      - Dữ liệu có phân phối đối xứng
      - Khi dữ liệu k có các giá trị ngoại lệ
      - Dữ liệu định lượng, cân nặng, chiều cao và có sự liên tục, ổn định
      - Khi cần tính toán các chỉ số khác dựa trên trung bình, mục tiêu là sử dụng toàn bộ thông tin của dữ liệu
      - khi dữ liệu sạch, kích thước mẫu lớnq



# 1.   Xử lý cột `last review`

Chuyển `last review` sang kiểu dữ liệu datetime và check thấy tới năm 2058 luôn

Nên làm với tất cả ngày tháng trong dữ liệu

In [ ]:
# rs = pd.to_datetime(df['last review'])  #được sử dụng để chuyển đổi kdl trong cột 'last review' sang kdl ngày tháng
# rs.min(), rs.max()  #in ra gtri ntns nhỏ nhất và lớn nhất của cột dữ liệu

#ta phải truyền trực tiếp vì biến rs chỉ là biến lưu trữ thôi chứ k cập nhật dữ liệu
df['last review'] = pd.to_datetime(df['last review'])
df['last review'].min(), df['last review'].max()

(Timestamp('2012-07-11 00:00:00'), Timestamp('2058-06-16 00:00:00'))

In [ ]:

year_value =df[df['last review'].apply(lambda x: x.year> 2022)] # tìm ra gtri những năm lớn hơn 2022
print(len(year_value)) # đếm số lượng dữ liệu lớn hơn năm 2022

# tìm gtri trung vị bằng hàm median
median_value = df['last review'].median()
print(median_value)

df.loc[year_value.index,'last review'] = median_value # thay thế các gtri này thành gtri trung vị của cột last review

5
2019-06-14 00:00:00


`last review` là cột chứa thông tin về ngày đánh giá gần nhất hoặc lần xem xét cuối cùng cho một sản phẩm, mục nào đó.

Cột này thường bao gồm các giá trị ngày tháng, cho biết thời điểm lần cuối cùng mà một đánh giá hoặc kiểm tra được thực hiện.

In [ ]:
# sau khi thay thế gtri có năm lớn hơn 2022 thì ta lại check những gtri null
# ta thấy những gtri null có nhiều lượt review ở cột 'number of reviews' nhưng sang cột 'last review' thì lại null
# nên ta thay thể những gtri null đó cũng thành gtri trung vị như trên

# tìm ra những gtri null
null_value = df[(df['last review'].isnull()) & (df['number of reviews']>0)]
print(len(null_value))

# thay thế những gtri null bằng gtri trung vị
df.loc[null_value.index, 'last_review'] = median_value

35


# 2. Xử lý cột `host_identity_verified` có giá trị Null

### Sự khác nhau giữa df [ df [ 'service ' ] ] và df [ 'service ' ]
- df [ 'service ' ] : là truy cập tới cả một cột trong df, cụ thể là service. Kết quả: trả về một Series chứa tất cả các giá trị của cột service đó.</br>
- df [ df [ 'service ' ] ] : là lọc tất cả các hàng trong cột service của df

In [ ]:
#ý tưởng tìm ra gtri null và thay thế chúng thành unconfirmed
# cach 1: (chỉ tìm dc gtri null)
# null_value_host = df[df['host_identity_verified'].isnull()]
# print(len(null_value_host))

#cách 2: (lấy dc full tất cả gtri)
all_value = df['host_identity_verified'].value_counts(dropna = False)
#value_counts() : đếm số lần xuất hiện của tất cả các gtri trong cột
#dropna = False : là đảm bảo giá trị null cũng dc tính và in ra
print(all_value)
print()

# thay thế gtri null thành unconfirmed
df['host_identity_verified'].fillna('unconfirmed', inplace = True)
#fillna() : làm đầy gtri rỗng bằng 1 gtri khác
#inplace = True : là thay đổi df ban đầu nếu False là tạo ra 1 df mới

# sau khi thay thế ta check lại
value_after = df['host_identity_verified'].value_counts(dropna = False)
print(value_after)

host_identity_verified
unconfirmed    51200
verified       51110
NaN              289
Name: count, dtype: int64

host_identity_verified
unconfirmed    51489
verified       51110
Name: count, dtype: int64


# 3.   Xử lý `minimum nights,availability 365`

In [ ]:
# trong 2 cột có những gtri âm vô lý nên mình sẽ nhân với -1 để chúng thành dương
df['minimum nights'] = df['minimum nights'].apply(lambda x : x*-1 if x<0 else x) # nếu x<0 (tức gtri âm) thì nhân với -1 còn else x( tức gtri dương) thì giữ nguyên
df['availability 365'] = df['availability 365'].apply(lambda x : x*-1 if x<0 else x)

Trong cột `minimun nights` có nhiều gtri quá lớn như 399,452.. mà trong khi `availability 365` max là 1 năm có 365 ngày. Ta thấy k hợp lệ,bên phía cột `availability 365` cũng có những gtri tương tự. Đây là các gtri ngoại lệ(outlier)

Khi loại bỏ ngoại lệ, điểm cutpoint được sử dụng để xác định các giá trị ngoại lệ hoặc giá trị không mong muốn. Những giá trị vượt quá điểm cutpoint thường bị coi là ngoại lệ và có thể bị loại bỏ hoặc xử lý theo cách khác.

In [ ]:
#loại bỏ dữ liệu outlier

cutpoint = df['minimum nights'].quantile(0.99) #tìm ra điểm cutpoint
# Xác định điểm cutpoint dựa trên phân vị 99%, các giá trị nằm ngoài khoảng này (cụ thể là lớn hơn điểm cutpoint) có thể được coi là ngoại lệ
print('Your cutpoint: ',cutpoint)

# Loại bỏ các giá trị lớn hơn điểm cutpoint
df = df[(df['minimum nights']<=cutpoint)]
#làm tương tự với cột availability 365

cutpoint_365 = df['availability 365'].quantile(0.99)
df = df[df['availability 365'] <= cutpoint]

Your cutpoint:  55.0


# 4.Xử lý cột `price` và `service fee`
Khi có những kí tự đb thì pd sẽ đọc thành string </br>
Vì cột có những kí tự "$" </br>

Và nhiều hàng gtri cao như : "$1,052",.... có chứa cả dấu " , " nữa ta cũng cần loại bỏ để biến chúng từ str thành int



In [ ]:
# loại bỏ những gtri nulll - vì null thì lấy đâu ra gtri mà biển đối sang string=)) hơn nữa hàm astype() k thể sd khi có gtri null
df = df[~df['price'].isnull()]  # dấu ~ là đảo ngược gtri, mặc định của isnull() là True đảo lại thành False
# khi thực hiện xong hàm thì cột df2 sẽ chứa tca những gtri null của cột price mà k hề ảnh hướng tới cột price

#k thể dùng như này vì hàm isnull() chỉ loại bỏ những gtri null trong 1 cột còn notnull() là loại bỏ hẳn cả 1 hàng, ảnh hưởng tới những cột khác trong df
# df['price'] = df[df['price'].notnull()]

# phải lọc ra trước và biến đổi thành '0' để đảm bảo đúng định dạng str cho các bước biến đổi sau chỉ cần .str là xong
#loại bỏ các gtri nan, NaN, nAn,... dưới dạng str thay thế thành '0'
df['price'] = df['price'].apply(lambda x: '0' if str(x).lower() == 'nan' else str(x)) #kiểm tra xem giá trị x có chuỗi 'nan' hay không, bất kể chữ hoa hay chữ thường

# loại bỏ những gtri rỗng thay thế thành '0'
df['price'] = df['price'].apply(lambda x: '0' if str(x).strip() == '' else str(x))
# strip() là loại bỏ các khoảng trắng ở đầu và cuối chuỗi. Ví dụ, ' 123 ' sẽ trở thành '123'.


# loại bỏ kí tự '$' và kí ',' trong những gtri str
# df['price'] = df['price'].apply(lambda x : str(x).replace('$','') if '$' in str(x) else str(x))
# df['price'] = df['price'].apply(lambda x : str(x).replace(',','') if ',' in x else x)
# hay
df['price'] = df['price'].str.replace('$', '')
df['price'] = df['price'].str.replace(',', '')

# biến đổi str thành int
df['price'] = df['price'].astype(int)


# tương tự với cột 'service fee'
df3 = df[~df['service fee'].isnull()]
df['service fee'] = df['service fee'].astype(str)
#check xem có gtri nào nan,Nan,NaN,nAn,.. hoặc rỗng không

df['service fee'] = df['service fee'].apply(lambda x: '0' if str(x).lower() == 'nan' else str(x))
df['service fee'] = df['service fee'].apply(lambda x: '0' if str(x).strip() == '' else str(x))
# df['service fee'] == df['service fee'].apply(lambda x : '0' if str(x).lower() in ['nan','none',''] else str(x))

df['service fee'] = df['service fee'].apply(lambda x : str(x).replace('$','') if '$' in str(x) else str(x))
#convert từ str sang int
df['service fee'] = df['service fee'].astype(int)
df.head(6)

<ipython-input-9-f481cfda9fd3>:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price'] = df['price'].apply(lambda x: '0' if str(x).lower() == 'nan' else str(x)) #kiểm tra xem giá trị x có chuỗi 'nan' hay không, bất kể chữ hoa hay chữ thường


id                                            NAME      host id  \
10   1006859              Cute & Cozy Lower East Side 1 bdrm   1280143094   
13   1008516  Lovely Room 1, Garden, Best Area, Legal rental  26802410424   
22   1013487                  CBG CtyBGd HelpsHaiti rm#1:1-4  53754237010   
23   1014039                        CBG Helps Haiti Room#2.5  87668933180   
42   1024533             Beautiful Sunny Park Slope Brooklyn  50357575975   
111  1062641             Times Square, Safe, Clean and Cozy!  61210263217   

    host_identity_verified host name neighbourhood group   neighbourhood  \
10                verified   Miranda           Manhattan       Chinatown   
13                verified     Darcy             brookln     South Slope   
22                verified      Mary            Brooklyn      Park Slope   
23             unconfirmed   William            Brooklyn      Park Slope   
42                verified     Alina            Brooklyn     South Slope   
111            unconfirmed     Kevin           Manhattan  Hell's Kitchen   

          lat      long        country  ... minimum nights number of reviews  \
10   40.71344 -73.99037  United States  ...            1.0             160.0   
13   40.66829 -73.98779  United States  ...            4.0             167.0   
22   40.68069 -73.97706  United States  ...            2.0             130.0   
23   40.67989 -73.97798  United States  ...            1.0              39.0   
42   40.66278 -73.97966  United States  ...            3.0              15.0   
111  40.75527 -73.99291  United States  ...            2.0             334.0   

    last review reviews per month  review rate number  \
10   2019-06-09              1.33                 3.0   
13   2019-06-24              1.34                 4.0   
22   2019-07-01              1.09                 4.0   
23   2019-01-01              0.37                 3.0   
42   2019-05-27              0.39                 2.0   
111  2019-06-16              3.00                 1.0   

     calculated host listings count  availability 365  \
10                              4.0               1.0   
13                              3.0              47.0   
22                              6.0              54.0   
23                              6.0               9.0   
42                              1.0              33.0   
111                             2.0              34.0   

                                           house_rules  license last_review  
10                                                 NaN      NaN         NaT  
13                                                 NaN      NaN         NaT  
22   Arrival time can be no later than 10:00PM. No ...      NaN         NaT  
23                                                 NaN      NaN         NaT  
42   No smoking inside. Since I have a dog, there a...      NaN         NaT  
111  You will be very comfortable in our home but w...      NaN         NaT  

[6 rows x 27 columns]

# **4.Thống kê dữ liệu**

## Neighbourhood group - barchart


In [ ]:
# Thống kê loại khu vực
print(df['neighbourhood'].value_counts())
print('-' * 25)
print(df['neighbourhood'].value_counts(normalize=True))


neighbourhood
Williamsburg          3964
Bedford-Stuyvesant    3096
Bushwick              2308
Harlem                2195
East Village          1698
                      ... 
Mariners Harbor          1
East Morrisania          1
Bay Terrace              1
Unionport                1
Woodrow                  1
Name: count, Length: 210, dtype: int64
-------------------------
neighbourhood
Williamsburg          0.098791
Bedford-Stuyvesant    0.077159
Bushwick              0.057520
Harlem                0.054704
East Village          0.042318
                        ...   
Mariners Harbor       0.000025
East Morrisania       0.000025
Bay Terrace           0.000025
Unionport             0.000025
Woodrow               0.000025
Name: proportion, Length: 210, dtype: float64


In [ ]:
import  matplotlib
import plotly.express as px

rm = df['neighbourhood'].value_counts()

# gtri bảng
fig = px.bar(y = rm.values, # tổng số lượng các loại khu vực
             x = rm.index, # các loại khu vực (name)
             color = rm.index,  #màu dựa theo index khu vực
             color_discrete_sequence=px.colors.sequential.Plasma, #blend màu
             text = rm.values,
             title = "Neighbourhoods",
             template = "plotly_dark"
             )

# thông số bảng, bố cục biểu đồ
fig.update_layout(
    xaxis_title = "Neighbourhood",
    yaxis_title = "count",
    font = dict(size=17,family="Franklin Gothic"),
    height=1000
    )

fig.show()

## Scatter Mapbox - biểu đồ phân tán

In [ ]:
print(df['neighbourhood group'].value_counts())  # số lượng khu vực lân cận

neighbourhood group
Brooklyn         17914
Manhattan        17374
Queens            4077
Bronx              679
Staten Island      178
brookln              1
Name: count, dtype: int64


In [ ]:
#so sánh price dựa theo từng khu vực lân cận

fig = px.scatter_mapbox(df,
           lat="lat",  #latitude - vĩ độ
           lon="long", #longitude - kinh độ
           opacity = 0.3,
           hover_name="neighbourhood group",
           hover_data=["neighbourhood group", "price"], #thông tin hiển thị gtri khi user di chuột trên bản đồ
           color="price", #màu dựa theo price
           color_discrete_sequence=px.colors.sequential.PuBuGn, # mã màu
           zoom=10
           )

fig.update_layout(mapbox_style="open-street-map") # cũng có thể sd kiểu bản đồ "carto-positron"
fig.update_layout(margin={"r":0,"t":0,"l":0,"b":0}, #loại bỏ lề xung quanh biểu đồ, tận dụng toàn bộ không gian
                  font = dict(size=17,family="Franklin Gothic"))
fig.show()

In [ ]:
print(df['room type'].value_counts())

room type
Entire home/apt    20456
Private room       19004
Shared room          743
Hotel room            22
Name: count, dtype: int64


## Piechart - Constrction Year

In [ ]:
print(df['Construction year'].value_counts())

Construction year
2009.0    2088
2019.0    2080
2014.0    2071
2012.0    2059
2006.0    2048
2020.0    2033
2007.0    2024
2022.0    2023
2008.0    2020
2018.0    2020
2016.0    2011
2021.0    2003
2005.0    1990
2015.0    1988
2003.0    1967
2017.0    1966
2010.0    1963
2004.0    1947
2013.0    1945
2011.0    1926
Name: count, dtype: int64


In [ ]:
constr_year = df["Construction year"].value_counts()
fig = px.pie(
    values = constr_year.values,
    names = constr_year.index,
    color_discrete_sequence = ['#CDEDDD', '#EF553B', '#00CC96', '#AB63FA'], #mã hex
    title= 'Construction Year',template='plotly_dark'
)
fig.update_layout(
    font=dict(size=20,family="Franklin Gothic"),
    height=900)
fig.show()

## Line Chart - Check identity by room type

In [ ]:
print(df['room type'].value_counts())
print('-' * 25)
print(df['host_identity_verified'].value_counts())

room type
Entire home/apt    20409
Private room       18956
Shared room          742
Hotel room            22
Name: count, dtype: int64
-------------------------
host_identity_verified
unconfirmed    20094
verified       20035
Name: count, dtype: int64


630.0


In [ ]:
# ktra xác minh danh tính theo từng kiểu phòng

# room_type = df['room type'].value_counts()
# room_price_avg = df['price'].median()
# print(room_price_avg)
# room = room_type.values
# print(df['price'])
# print(room_type)
constr_year = df['Construction year'].value_counts().sort_index()
fig = px.line( x=constr_year.index,
              y= constr_year.values,
              markers=True,
              line_shape='linear',
              text = constr_year.values,
              height=900)

fig.update_layout(title='Construction_Year',
                  xaxis_title='Count',
                  yaxis_title='Year',
                  font = dict(size=17, family="Franklin Gothic"),
                  )
fig.show()